### PR TITLE
styra-link: correct password flow

### PR DESCRIPTION
**Pull request by msorens**
### What code changed, and why?

Several issues revolving around the password entry flow when installing the Styra CLI from the plugin.
DAS-770, DAS-771, DAS-772.

### Definition of done

OLD:
image

NEW:
image

1. fix poor error reporting upon bad password.

2. allow password retries without re-downloading full binary: Add loop for retry but move the binary fetch outside the loop 

3. cancel out of password entry closes progress bar.

### How to test

You can install either from source code or from GitHub.

#### Install from source

1. Switch to current branch.
2. Run `npm install`.
3. Open project in VSCode.
4. In the debugger be sure you have "Run Extension" mode selected.
5. Invoke "run" (<kbd>F5</kbd>).

#### Install from GitHub

1. Go to the latest release on the [release page](https://github.com/StyraInc/vscode-styra/releases).
2. Download the additional version piggybacked into that release, labelled `vscode-styra-2.0.1-next.4.vsix`.
3. Install via the standard command: `code --install-extension vscode-styra-2.0.1-next.4.vsix`

### CODE REVIEW

File: src/lib/styra-install.ts [hunk 1]
```diff
@@ -45,12 +45,12 @@ export class StyraInstall {
   }
 
   static async checkCliInstallation(): Promise<boolean> {
-    if (await StyraInstall.styraCmdExists()) {
+    if (await this.styraCmdExists()) {
```
Comment (msorens) [inline src/lib/styra-install.ts:48]: Had a mix of class name vs. `this`. Making them all consistent.

File: src/lib/styra-install.ts [hunk 2]
```diff
@@ -60,15 +60,29 @@ export class StyraInstall {
       `Styra CLI ${description}. Would you like to install it now?`, 'Install');
 
     if (selection === 'Install') {
+      const tempFile = path.join(os.homedir(), this.BinaryFile);
       info('Installing Styra CLI. This may take a few minutes...');
       try {
-        await this.installStyra();
```
Comment (msorens) [inline src/lib/styra-install.ts:65]: Separating `installStyra` into `downloadBinary` and `installOnPath`.

File: src/lib/styra-install.ts [hunk 3]
```diff
@@ -261,7 +278,7 @@ export class StyraInstall {
       value: state.pwd ?? '',
       prompt: `Enter admin password to install into ${STD_LINUX_INSTALL_DIR}`,
       validate: validateNoop,
-      shouldResume
+      shouldResume // TODO: override and delete temp file
```
Comment (msorens) [inline src/lib/styra-install.ts:281]: for later; low priority

File: src/test-jest/lib/styra-install.test.ts [hunk 4]
```diff
@@ -33,9 +33,13 @@ describe('StyraInstall', () => {
...
   const spy = new OutputPaneSpy();
 
+  function setPrivateMock(functionName: string, mock: jest.Mock) {
+    (StyraInstall as any)[functionName] = mock;
+  }
+
   beforeEach(() => {
-    // eslint-disable-next-line dot-notation
-    StyraInstall['installStyra'] = jest.fn().mockResolvedValue('');
+    setPrivateMock('downloadBinary', jest.fn().mockResolvedValue(''));
```
Comment (msorens) [inline src/test-jest/lib/styra-install.test.ts:41]: Introduced `setPrivateMock` to get rid of the smattering of eslint disablements.

**Review by srenatus — APPROVED** [review]

File: src/lib/styra-install.ts [hunk 5]
```diff
@@ -178,22 +191,26 @@ export class StyraInstall {
...
   }
 
+  private static async installOnPath(tempFileLocation: string) {
+    if (this.isWindows()) {
+      await moveFile(tempFileLocation, this.ExeFile);
+      await this.adjustWindowsPath(this.ExePath);
+    } else {
+      const state = await this.collectInputs();
+      // see https://stackoverflow.com/q/39785436/115690 for ideas on running sudo
+      const args = ['-c', `echo ${state.pwd} | sudo -S bash -c 'mv ${tempFileLocation} ${this.ExeFile}'`];
```
[the remark this answers is not in the document]
Comment (msorens) [reply src/lib/styra-install.ts:204]: I certainly agree with your point but I am not sure there is a way to give a user a shell prompt from within a VSCode command. I like the idea though, so will put it on my backlog but merge as is for now. Thanks!

[the remark this answers is not in the document]
Comment (srenatus) [reply src/lib/styra-install.ts:204]: The only winning move is not to play -- we can install the binary anywhere, and use it from anywhere. So we can use a location where an unprivileged user can write to. `.styra/bin/` or something. We'll add a note saying that they need to add that to their PATH to have it be available on the command line. If they only interact with the link tool using the VSCode plugin, there's no need for updating the $PATH even.

**Pull request merged**